In [1]:
import pandas as pd

In [2]:
data_dir = "./data/raw_data/"
save_dir = "./data/preprocessed_data/"

Read in the data

In [3]:
cv_table = pd.read_csv(data_dir + "cv_table.csv")
ndr_table = pd.read_csv(data_dir + "ndr_table.csv")
poll_table = pd.read_csv(data_dir + "poll_table.csv")

latlong_cv = pd.read_csv(data_dir + "latlong_cv.csv")
latlong_ndr = pd.read_csv(data_dir + "latlong_ndr.csv")
latlong_pollination = pd.read_csv(data_dir + "latlong_pollination.csv")


We have three tables of data, one for each contribution of nature. We also have three tables relating each data point to a longitude and a latitude.

Let's look at the tables first!

# cv_table

In [4]:
cv_table.describe(include = 'all')

fid         UN_cur        UN_ssp1        UN_ssp3  \
count  686665.000000  686665.000000  686665.000000  686665.000000   
mean   343333.000000       2.482690       2.822749       2.982682   
std    198223.255633       0.511879       0.573237       0.618028   
min         1.000000       1.030057       1.237034       1.297790   
25%    171667.000000       2.082644       2.376177       2.492883   
50%    343333.000000       2.440013       2.777381       2.935599   
75%    514999.000000       2.817269       3.259844       3.419952   
max    686665.000000       4.291871       4.817462       5.000000   

             UN_ssp5        pop_cur       pop_ssp1       pop_ssp3  \
count  686665.000000  686665.000000  686665.000000  686665.000000   
mean        3.139066     485.270589     586.541642     605.606714   
std         0.644773    2772.848940    3590.280849    3368.830041   
min         1.311983       0.000000       0.000000       0.000000   
25%         2.667168       0.510938       0.423553       0.455772   
50%         3.107233      21.314615      21.012911      26.209813   
75%         3.619904     156.068106     153.478002     184.031578   
max         5.000000  236893.117288  270175.660403  247236.351206   

            pop_ssp5  
count  686665.000000  
mean      608.113528  
std      3719.613422  
min         0.000000  
25%         0.445830  
50%        21.459414  
75%       159.716874  
max    270048.141080

This table describes the coastal risk mitigation by natural habitats.
Some explanation of the table follow:

* **fid** - the id of the geographical location that the data describes. This id can be converted to lon/lat using the latlong table.  
* **UN_cur** - the **current** (2015) unmet need of the coastal risk mitigation in this particular area. If this figure is negative then it means that we have enough, and even a surplus, in this particular area. 
* **UN_ssp1** the unmet need of the coastal risk mitigation in the year 2050, given that scenario **ssp1** is realized. 
* **ssp1, ssp3, ssp5** - the three future scenarios. It is not perfectly clear which scenraio is which here. ssp1 is most likely the sustainability scenario, purely based on the observation that the unmet need in this column is lower than for the other scenarios. The other two scenarios are fossil-fuels and regional rivalry.
* **pop_cur** - the **current** (2015) number of people in this geographic region who are affected by coastal risk mitigation. If the unmet need is high and the population is high then a lot of people are negatively affected.
* **pop_sspX** - same as the above, but in the year 2050, for the three different future scenarios.

# ndr_table

In [5]:
ndr_table.describe(include = 'all')

fid     country   region        UN_cur       UN_ssp1  \
count   25243.000000       25243    25243  2.524300e+04  2.524300e+04   
unique           NaN         228        8           NaN           NaN   
top              NaN  Antarctica  UNKNOWN           NaN           NaN   
freq             NaN        6511     6557           NaN           NaN   
mean    31136.198669         NaN      NaN  2.847848e+06  2.982161e+06   
std     20508.712007         NaN      NaN  8.093390e+06  9.765498e+06   
min      2262.000000         NaN      NaN  0.000000e+00  0.000000e+00   
25%     13083.500000         NaN      NaN  0.000000e+00  0.000000e+00   
50%     25299.000000         NaN      NaN  2.304164e+03  7.343517e+02   
75%     56810.500000         NaN      NaN  2.418531e+06  1.757585e+06   
max     64799.000000         NaN      NaN  1.714816e+08  3.159408e+08   

             UN_ssp3       UN_ssp5     rurpopcur    rurpopssp1    rurpopssp3  \
count   2.524300e+04  2.524300e+04  2.524300e+04  2.524300e+04  2.524300e+04   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    3.464451e+06  2.888546e+06  1.693007e+05  1.008505e+05  2.328594e+05   
std     1.072992e+07  8.570622e+06  6.919647e+05  4.071316e+05  9.817925e+05   
min     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%     7.217091e+02  7.175778e+02  0.000000e+00  0.000000e+00  0.000000e+00   
75%     2.092982e+06  1.920021e+06  4.479987e+04  2.482487e+04  4.230556e+04   
max     2.093042e+08  2.064157e+08  1.667823e+07  8.933015e+06  2.162865e+07   

          rurpopssp5  
count   2.524300e+04  
unique           NaN  
top              NaN  
freq             NaN  
mean    1.003789e+05  
std     4.030947e+05  
min     0.000000e+00  
25%     0.000000e+00  
50%     0.000000e+00  
75%     2.592468e+04  
max     8.607843e+06

This table describes the improvement of water quality by filtering of pollutants (in our case probably excessive amounts of nutrients, making the water quality bad) when water passes by natural habitats such as forests and wetlands.

The columns are the same as in the previous table, although `pop` is replaced by `rurpop`. Charlotte has confirmed that this is still the same metric though.

We also have two extra columns `country` and `region`. We will most probably not use these, especially as we don't actually have them for each table.

One thing we can immediately observe is that the count (number of entries) is much lower, the previous table had 686665 entries while this one has only 25243. This needs to be investigated. Does this mean that we don't have data for the whole earth for this particular contribution of nature?

# poll_table

In [6]:
poll_table.describe(include = 'all')

fid        UN_cur       UN_ssp1       UN_ssp3       UN_ssp5  \
count   9160.000000  9.160000e+03  9.160000e+03  9.160000e+03  9.160000e+03   
mean   25270.656441  9.630618e+04  8.972863e+04  9.375806e+04  9.990251e+04   
std    10032.053267  4.348216e+05  4.136170e+05  4.113846e+05  4.335244e+05   
min     8845.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    16672.000000  2.088596e+00  2.046068e+00  9.315915e+00  9.550000e+00   
50%    22971.500000  1.027879e+03  9.781146e+02  1.492830e+03  1.594218e+03   
75%    33446.250000  3.107308e+04  2.827017e+04  3.439839e+04  3.474784e+04   
max    51951.000000  2.322062e+07  2.249534e+07  2.321814e+07  2.270223e+07   

            pop_cur      pop_ssp1      pop_ssp3      pop_ssp5  
count  9.160000e+03  9.160000e+03  9.160000e+03  9.160000e+03  
mean   2.574799e+05  3.993871e+05  4.659653e+05  3.603159e+05  
std    1.004224e+06  1.546353e+06  1.681148e+06  1.450151e+06  
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
50%    6.223414e+02  3.940410e+03  3.460522e+03  0.000000e+00  
75%    1.263994e+05  1.874406e+05  2.276243e+05  1.468894e+05  
max    2.338777e+07  3.778232e+07  3.282686e+07  3.222564e+07

This table describes the pollination of crops coming from natural habitats inhabiting pollinators (ex. bees).

The columns are the same as previously.

Here though we have an even lower count. What does this mean? A hypothesis could be that the resolution is simply not just not as high for the data here.

# latlong_cv

In [7]:
latlong_cv.describe(include = 'all')

fid            lat            lng
count  686665.000000  686665.000000  686665.000000
mean   343333.000000       9.346826      23.706301
std    198223.255633      32.202816     102.750279
min         1.000000     -58.480000    -180.000000
25%    171667.000000     -10.600000     -73.070000
50%    343333.000000       9.600000      12.640000
75%    514999.000000      35.450000     123.250000
max    686665.000000      62.800000     180.000000

This looks good, just a simple mapping from the **fid** to the coordinates. We have the same number of mappings as we have entrys in the `cv_table`.

# latlong_ndr

In [8]:
latlong_ndr.describe(include = 'all')

fid           lat           lng
count  64800.000000  64800.000000  64800.000000
mean   32399.500000      0.000000      0.000000
std    18706.293059     51.961123    103.923449
min        0.000000    -89.500000   -179.500000
25%    16199.750000    -44.750000    -89.750000
50%    32399.500000      0.000000      0.000000
75%    48599.250000     44.750000     89.750000
max    64799.000000     89.500000    179.500000

Here the number of entries is actually higher than the number of entries in the `ndr_table`. This is a bit strange, what does this mean?

Moreover, why do we even need to have different mappings for the different contributions of nature? Couldn't they just use the same mappings between id and coordinates? The answer to this is probably that, as suggested before, the world map resolutions of the nature's contributions are different, with coastal risk having the most detailed data.

# latlong_pollination

In [9]:
latlong_pollination.describe(include = 'all')

fid           lat           lng
count  64800.000000  64800.000000  64800.000000
mean   32399.500000      0.000000      0.000000
std    18706.293059     51.961123    103.923449
min        0.000000    -89.500000   -179.500000
25%    16199.750000    -44.750000    -89.750000
50%    32399.500000      0.000000      0.000000
75%    48599.250000     44.750000     89.750000
max    64799.000000     89.500000    179.500000

Once again the same thing, the number of entries are the same as in the `latlong_ndr` table, 64800, while the number of entries in the `poll_table` is just 9160...



# Actual data

We also take a look at the actual data, just to get a quick feel for the actual figures:

In [10]:
cv_table.head()

fid    UN_cur   UN_ssp1   UN_ssp3   UN_ssp5  pop_cur  pop_ssp1  pop_ssp3  \
0    1  3.282099  3.684031  3.684031  4.135186      0.0       0.0       0.0   
1    2  2.924018  3.282099  3.282099  3.684031      0.0       0.0       0.0   
2    3  2.924018  3.282099  3.282099  3.684031      0.0       0.0       0.0   
3    4  2.924018  3.282099  3.282099  3.684031      0.0       0.0       0.0   
4    5  2.924018  3.282099  3.282099  3.684031      0.0       0.0       0.0   

   pop_ssp5  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0

Nothing surprising here! The id:s start from 1 and go up. We see an unmet need for the first 5 geographical regions both currently and for the 3 different future scenarios. We also see that nobody in this region actually gets directly affected by any changes to the natural habitats mitigating coastal risk. This could be because nobody lives there, or simply perhaps because there is no coast in this area.

Let's look at the `ndr_table`.

In [11]:
ndr_table.sample(5)

fid     country         region       UN_cur      UN_ssp1  \
1816    7049      Norway        Eurasia       0.0000       0.0000   
4906   10881      Canada  North America  251095.4531  282982.1563   
18782  56084  Antarctica        UNKNOWN       0.0000       0.0000   
11137  21862       China     North Asia  977346.8750  506788.3125   
21197  60446  Antarctica        UNKNOWN       0.0000       0.0000   

           UN_ssp3      UN_ssp5    rurpopcur   rurpopssp1   rurpopssp3  \
1816        0.0000       0.0000       0.0000       0.0000       0.0000   
4906   282843.2813  282331.4063       0.0000       0.0000       0.0000   
18782       0.0000       0.0000       0.0000       0.0000       0.0000   
11137  904785.9375  706997.0625  133477.7969  138011.8125  192777.6094   
21197       0.0000       0.0000       0.0000       0.0000       0.0000   

        rurpopssp5  
1816        0.0000  
4906        0.0000  
18782       0.0000  
11137  136074.8125  
21197       0.0000

We note that the **fid** goes above the maximum number of entries in the table (which was 25243). 

Same for `poll_table`:

In [12]:
poll_table.sample(5)

fid         UN_cur        UN_ssp1        UN_ssp3        UN_ssp5  \
7502  36106       0.000000       0.000000       0.000000       0.000000   
5952  29086  157730.522600  153079.427500  156874.474800  150008.988400   
3452  19358       7.015983       0.236388       0.008241       4.984732   
9070  47266   16661.009490   17251.099750   21595.842690   18808.988670   
3197  18809    3308.591392    2517.527436    6648.825732    3076.228823   

          pop_cur     pop_ssp1     pop_ssp3     pop_ssp5  
7502  27905.13281  25810.31641  44168.75000  22889.46680  
5952      0.00000      0.00000      0.00000      0.00000  
3452  16856.60352  12791.45605  15621.77637  12792.88965  
9070      0.00000      0.00000      0.00000      0.00000  
3197      0.00000      0.00000      0.00000      0.00000

And for an example lat_long-table let's check out `latlong_cv`:

In [13]:
latlong_cv.sample(5)

fid    lat     lng
535596  535597  39.58  125.39
322120  322121   5.39  103.10
624221  624222  53.88    8.67
274125  274126  -0.76    8.93
308132  308133   2.84  117.63

# Preprocessing

The next step is to perform the preprocessing (if needed) and save the data in new files. 
This might need to be done in several iteration as we don't know at the moment exactly how we will use the data. One thing that will be good to have done already is to join the lat/lon tables with the data tables.

## Join lat/lon tables with the data tables

Start with `cv_table`:

In [14]:
cv_table_out = pd.merge(cv_table, latlong_cv, on='fid')
cv_table_out.sample()

fid    UN_cur   UN_ssp1   UN_ssp3   UN_ssp5  pop_cur  pop_ssp1  \
43819  43820  2.236068  2.685377  2.924018  2.924018      0.0       0.0   

       pop_ssp3  pop_ssp5    lat    lng  
43819       0.0       0.0 -46.84  37.76

`fid` column is not needed, drop this.

In [15]:
cv_table_out.drop(columns=['fid'], inplace=True)
cv_table_out.sample()

UN_cur   UN_ssp1   UN_ssp3   UN_ssp5       pop_cur      pop_ssp1  \
383401  2.667168  3.203101  3.360421  3.487751  17303.074911  22832.171105   

            pop_ssp3      pop_ssp5    lat     lng  
383401  34624.397664  22447.789138  12.36  123.63

Looks good! Let's do the same for `ndr_table` and `poll_table`:

In [16]:
ndr_table_out = pd.merge(ndr_table, latlong_ndr, on ='fid')

# Remove the country and region columns, which we will probably not use, and rename columns
ndr_table_out.drop(columns=['fid', 'country', 'region'], inplace=True) 
ndr_table_out.rename(columns={"rurpopcur": "pop_cur", 
                              "rurpopssp1": "pop_ssp1", 
                              "rurpopssp3": "pop_ssp3",
                              "rurpopssp5": "pop_ssp5"}, inplace=True)

ndr_table_out.sample()

UN_cur  UN_ssp1  UN_ssp3  UN_ssp5  pop_cur  pop_ssp1  pop_ssp3  \
23393     0.0      0.0      0.0      0.0      0.0       0.0       0.0   

       pop_ssp5   lat    lng  
23393       0.0 -84.5  130.5

In [17]:

poll_table_out = pd.merge(poll_table, latlong_pollination, on='fid')
poll_table_out.drop(columns=['fid'], inplace=True)

poll_table_out.sample()

UN_cur     UN_ssp1     UN_ssp3     UN_ssp5  pop_cur  pop_ssp1  \
7412  580.534359  254.842453  566.289754  566.854007      0.0       0.0   

      pop_ssp3  pop_ssp5  lat    lng  
7412       0.0       0.0 -8.5  146.5

## Dropping rows without any inhabitants
We are interested in the number of affected people. Therefore, we can drop the data rows where there the population is 0 (currently, and for all of the 3 scenarios in 2050), e.g. places which are inhabited.

In [18]:
cols = ['pop_cur', 'pop_ssp1', 'pop_ssp3', 'pop_ssp5']
print("Number of rows before: " + str(cv_table_out.shape[0]))
cv_table_out = cv_table_out.loc[(cv_table_out[cols]!=0).any(axis=1)]
print("Number of rows after: " + str(cv_table_out.shape[0]))

Number of rows before: 686665
Number of rows after: 636425


Doing the same for the ndr_table and the poll table:

In [19]:
print("Number of rows in ndr_table before: " + str(ndr_table_out.shape[0]))
ndr_table_out = ndr_table_out.loc[(ndr_table_out[cols]!=0).any(axis=1)]
print("Number of rows in ndr_table after: " + str(ndr_table_out.shape[0]))

print("")

print("Number of rows in poll_table before: " + str(poll_table_out.shape[0]))
poll_table_out = poll_table_out.loc[(poll_table_out[cols]!=0).any(axis=1)]
print("Number of rows in poll_table after: " + str(poll_table_out.shape[0]))

Number of rows in ndr_table before: 25243
Number of rows in ndr_table after: 11913

Number of rows in poll_table before: 9160
Number of rows in poll_table after: 5130


A lot of data is lost doing this, especially proportionately in the case of the ndr and the poll datasets.

This is fine though since, as stated before, we are only interested in the number of affected people.

Now we are ready to save the files!

# Saving data

Doing a final reset of the indexes:

In [20]:
cv_table_out.reset_index(drop=True, inplace=True)
ndr_table_out.reset_index(drop=True, inplace=True)
poll_table_out.reset_index(drop=True, inplace=True)

In [21]:
cv_table_out.to_csv(save_dir + "cv_table_preprocessed.csv", index=False)
ndr_table_out.to_csv(save_dir + "ndr_table_preprocessed.csv", index=False)
poll_table_out.to_csv(save_dir + "poll_table_preprocessed.csv", index=False)